staging/src/kubevirt.io/client-go/api/v1/types.go

```diff
@@ -1334,14 +1351,21 @@ type SMBiosConfiguration struct {
 // MigrationConfiguration holds migration options
 // +k8s:openapi-gen=true
 type MigrationConfiguration struct {
-	AllowAutoConverge                 bool               `json:"allowAutoConverge,string"`
-	BandwidthPerMigration             *resource.Quantity `json:"bandwidthPerMigration,omitempty"`
-	CompletionTimeoutPerGiB           *int64             `json:"completionTimeoutPerGiB,string,omitempty"`
-	NodeDrainTaintKey                 *string            `json:"nodeDrainTaintKey,omitempty"`
-	ParallelOutboundMigrationsPerNode *uint32            `json:"parallelOutboundMigrationsPerNode,string,omitempty"`
-	ParallelMigrationsPerCluster      *uint32            `json:"parallelMigrationsPerCluster,string,omitempty"`
-	ProgressTimeout                   *int64             `json:"progressTimeout,string,omitempty"`
-	UnsafeMigrationOverride           bool               `json:"unsafeMigrationOverride,string"`
+	NodeDrainTaintKey                 *string `json:"nodeDrainTaintKey,omitempty"`
+	ParallelOutboundMigrationsPerNode *uint32 `json:"parallelOutboundMigrationsPerNode,string,omitempty"`
+	ParallelMigrationsPerCluster      *uint32 `json:"parallelMigrationsPerCluster,string,omitempty"`
+	*PerVMIMigrationConfiguration     `json:",inline"`
+}
+
+// PerVMIMigrationConfiguration holds migration options that can be configurated per vmi
+// +k8s:openapi-gen=true
+type PerVMIMigrationConfiguration struct {
```

Comment:
Is there a reason why we now want to expose all these other tunables per vmi?

Reply:
My thought was that if we assume that the user creating the VMI knows the most information about how the migration should be handled (i.e using post copy) it would also make sense for them to be able to configure the other migration options for their specific VM. Maybe that's giving them to much power though and they should only have the option between pre vs post copy. I don't have a strong opinion

Reply:
@vladikr @rmohr what do you think here?

I remember when we were first talking about migrations, we considered all these tunables to be things only the administrator would have the ability to modify. and that the role of the user would have all this abstracted away.

If that's still our thinking, then the post/pre copy setting itself might not even belong on the the vmi api... however, this could be useful.

Maybe this isn't a discussion we should have during this PR and the VMI api should stay as it is today with no per VMI migration tunings. We can introduce postcopy at the cluster config level still and have separate discussion focused on what tunings belong on the VMI itself?

Thoughts?

Reply:
I agree. I also don't think we should expose most of these tunables to the user.
Perhaps it's fine to expose `AllowAutoConverge` and `UsePostCopy` - but I'm also fine with keeping it in the hands of the admin.

Reply:
IMO one interesting question would be to ask: How can admin change migration options per vmi basis? (If that makes sense)

Reply:
Most of the tunables in this section make sense only as a cluster wide setting. The only two I would consider being set per migration are auto-converge and post-copy.
Perhaps we should split this tunable into an admin setting (allow/forbid auto-converge/post-copy) and a user per migration setting (forceAutoConverge and usePostCopy)

What do you think?

Reply:
yes, that might make sense.

For the sake of getting this feature merged, I suggest we keep everything at the admin level for now, and discuss the VMI API changes as a follow up. I don't think anyone questions the usefulness of post copy, which should make getting the fundamental logic in pretty straight forward.  I do think there are likely "pre vmi" migration tunables that can be exposed, but that likely goes beyond the scope of this initial feature implementation.